### PR TITLE
NNGraph: graph-order shapes with storage bridge (PyTorch-aligned API)

### PR DESCRIPTION
**Pull request by Muxas**
## Summary

Migrates the NNGraph stack to **graph-order shape labels** (outermost dimension first, aligned with PyTorch/HuggingFace conventions) while keeping tile/tensor storage unchanged. Conversion between graph and storage labels is metadata-only via `nn/shape_layout.hh` — no payload reordering.

### Core API

- Add `shape_layout.hh`: `graph_shape_to_storage`, `storage_shape_to_graph`, `graph_axis_to_storage`, `storage_axis_to_graph`.
- `NNGraph::TensorNode::shape()` returns graph labels; underlying `tensor::*` nodes keep storage labels.
- `NNGraph::tensor(shape)` accepts graph shapes and allocates storage with reversed labels.
- Rename runtime layout helpers to `storage_*` (drop Fortran/row-major wording).

### NN ops and modules

- Migrate NN ops (add, gemm, embedding, concat, layer_norm, rms_norm, cross_entropy, sdpa, rope, fiber/slice ops, etc.) to graph axes/shapes with storage conversion at the tensor lowering boundary.
- **GEMM:** `gemm(a, b, trans_a, trans_b, ...)` defaults to **no transpose** (`trans_a=false`, `trans_b=false`), i.e. `y = alpha * a @ b` on graph shapes. Callers that need a transposed operand pass flags explicitly (e.g. `Linear` uses `trans_b=true` for weight `[out, in]`). Lowering swaps operands/flags when calling `tensor::gemm`.
- **Embedding:** graph-axis output shapes built in `nn/ops/embedding.cc`; tensor embedding keeps axis-aware validation/shape insertion.
- **Linear / MLP:** add `with_bias` to `module::Mlp`; wire GPT-2 and GPT-NeoX MLP biases into `Linear::forward` (no manual `add_fiber` in model code). Restore model `forward()` methods for tensor naming (`fc1_out`, `output` / `mlp_out`).

### Models, examples, and tests

- Update model constructors/forwards and shape comments for GPT-2, GPT-NeoX, GPT-Neo, LLaMA, BERT, RoBERTa, T5.
- Add `nntile/examples/graph_bind.py`; migrate C++/Python generate/training examples off `fortran_order` / `ravel("F")`.
- Refresh model test data generators, safetensors layout tests, and PyTorch parity tests for graph shapes.
- Drop layout-order terminology across NNGraph docs, tests, and comments.

### Intentionally out of scope (left at `graph_api`)

- **kernel / starpu / core / tile** layers — comment-only layout renames were reverted; these layers are not updated yet.
- **infra/gateway** — `model_loader.py` and gateway tests reverted; infra work is separate.

## Test plan

- [x] NN op tests (`tests_graph_nn_ops_*`)
- [x] Module tests (`tests_graph_module_*`, including linear, mlp, embedding)
- [x] Model tests (GPT-2, GPT-NeoX, GPT-Neo, LLaMA, BERT, RoBERTa, T5 where fixtures exist)
- [x] `tests_graph_model_safetensors_nntile_layout`
- [x] `python/tests/test_causal_mask.py`

<div><a href="https://cursor.com/agents/bc-39d0e41a-dff5-4eec-8d37-1809d15b2ca9"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/open-in-web-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/open-in-web-light.png">Open in Web</picture></a>&nbsp;<a href="https://cursor.com/background-agent?bcId=bc-39d0e41a-dff5-4eec-8d37-1809d15b2ca9"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/open-in-cursor-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/open-in-cursor-light.png">Open in Cursor</picture></a>&nbsp;</div>

### CODE REVIEW

File: nntile/include/nntile/model/bert/bert_self_attention.hh [hunk 1]
```diff
@@ -45,7 +45,9 @@ public:
     NNGraph::TensorNode* forward(
         NNGraph::TensorNode* x,
         NNGraph::TensorNode* mask = nullptr,
-        bool causal = false);
+        bool causal = false,
+        NNGraph::TensorNode* w_dense = nullptr,
```
Comment (Muxas) [inline nntile/include/nntile/model/bert/bert_self_attention.hh]: I do not understand why parameters of the layer appear as arguments to the forward. Parameters must be taken from the model or layer itself!

File: nntile/include/nntile/model/bert/bert_self_output.hh [hunk 2]
```diff
@@ -40,9 +40,12 @@ public:
                    DataType dtype = DataType::FP32);
 
     NNGraph::TensorNode* forward(
-        NNGraph::TensorNode* attn_heads,
+        NNGraph::TensorNode* dense_out,
         NNGraph::TensorNode* residual);
 
+    NNGraph::TensorNode* w_dense() const { return w_dense_; }
```
Comment (Muxas) [inline nntile/include/nntile/model/bert/bert_self_output.hh]: I do not understand why parameters of the layer appear as arguments to the forward. Parameters must be taken from the model or layer itself!

File: nntile/include/nntile/model/llama/llama_attention.hh [hunk 3]
```diff
@@ -63,18 +61,18 @@ public:
...
+    //! @param cos RoPE cos tensor (batch, seq, head_size/2), may be nullptr to
     //! skip RoPE
     //! @param mask Optional BOOL attention mask (k_seq, q_seq), may be
-    //! nullptr. Build with ``nntile::sdpa_causal_mask_bool_fortran_fill``
+    //! nullptr. Build with ``nntile::sdpa_causal_mask_bool_fill``
     //! for causal
     //! LM.
-    //! @param k_cache Optional KV cache for K (head_size, max_seq, batch, n_head_kv)
-    //! @param v_cache Optional KV cache for V (head_size, max_seq, batch, n_head_kv)
+    //! @param k_cache Optional KV cache for K (batch, max_seq, n_head_kv, head_size)
```
Comment (Muxas) [inline nntile/include/nntile/model/llama/llama_attention.hh]: full shape reversion means (head_size, max_seq, batch, n_head_kv) shall become (n_head_kv, batch, max_seq, head_size). I understand, that this is strange, as batch shall be outermost axis, but I am sure we use batch=1 and the dimension itself can appear anywhere without harming numerics.

File: nntile/src/model/bert/bert_attention.cc [hunk 4]
```diff
@@ -34,9 +34,10 @@ NNGraph::TensorNode* BertAttention::forward(
     NNGraph::TensorNode* mask,
     bool causal)
 {
-    NNGraph::TensorNode* heads =
-        self_attn_.forward(x, mask, causal);
-    return self_out_.forward(heads, x);
+    NNGraph::TensorNode* dense =
+        self_attn_.forward(x, mask, causal,
+            self_out_.w_dense(), self_out_.b_dense());
```
Comment (Muxas) [inline nntile/src/model/bert/bert_attention.cc]: Why do we need to provide here weight and bias from other layer?

File: nntile/src/model/bert/bert_embeddings.cc [hunk 5]
```diff
@@ -1,8 +1,12 @@
+#include <nntile/common.hh>
```
Comment (Muxas) [inline nntile/src/model/bert/bert_embeddings.cc]: Legal header first, all includes after -- follow the style guide!

File: nntile/src/model/bert/bert_self_attention.cc [hunk 6]
```diff
@@ -36,35 +39,37 @@ BertSelfAttention::BertSelfAttention(NNGraph* graph,
...
     v_bias_->set_name(tensor_name("v_bias"));
     register_parameter("v_bias", v_bias_);
 }
 
 NNGraph::TensorNode* BertSelfAttention::forward(
     NNGraph::TensorNode* x,
     NNGraph::TensorNode* mask,
-    bool causal)
+    bool causal,
+    NNGraph::TensorNode* w_dense,
```
Comment (Muxas) [inline nntile/src/model/bert/bert_self_attention.cc]: weight and bias shall come from the layer itself!

File: nntile/src/model/gpt2/gpt2_model.cc [hunk 7]
```diff
@@ -71,10 +71,10 @@ NNGraph::TensorNode* Gpt2Model::forward(
 
     NNGraph::TensorNode* wte_out = wte_.forward(input_ids);
     NNGraph::TensorNode* wpe_out = wpe_.forward(position_ids);
-    NNGraph::TensorNode* embed =
-        add(1.0, wte_out, 1.0, wpe_out);
     NNGraph::TensorNode* x =
-        transpose(embed, 2);
+        add(1.0, wte_out, 1.0, wpe_out);
+    // Embedding C-order [hidden, batch, seq] -> [batch, seq, hidden].
+    x = transpose(x, 2);
```
Comment (Muxas) [inline nntile/src/model/gpt2/gpt2_model.cc]: I find this transposition to be very strange! Do we really need it? Why not work in [batch, seq, hidden] shape immediately? Seems like some operation (gemm, probably) moves batch into internal axis.

**Review by Muxas — COMMENTED** [review]
Partial review.

File: nntile/src/model/gptneox/gptneox_mlp.cc [hunk 8]
```diff
@@ -44,16 +43,14 @@ GptneoxMlp::GptneoxMlp(NNGraph* graph,
...
     NNGraph::TensorNode* input)
 {
-    NNGraph::TensorNode* w1 = fc1().weight_tensor();
-    NNGraph::TensorNode* hidden =
-        gemm(w1, input, 1.0, true, false, 1, 0);
+    NNGraph::TensorNode* hidden = fc1().forward(input);
     hidden->set_name(tensor_name("fc1_out"));
-    hidden = add_fiber(1.0, fc1_bias_, 1.0, hidden, 0, 0);
+    const Index hidden_feature_axis = hidden->ndim() - 1;
+    hidden = add_fiber(1.0, fc1_bias_, 1.0, hidden, hidden_feature_axis, 0);
```
Comment (Muxas) [inline nntile/src/model/gptneox/gptneox_mlp.cc]: I think such add_fiber calls shall be inside forward operation of the linear layer (fc1() in this case).

File: nntile/src/model/gptneox/gptneox_model.cc [hunk 9]
```diff
@@ -59,8 +59,8 @@ NNGraph::TensorNode* GptneoxModel::forward(
             "GptneoxModel::forward: input_ids must be non-null");
     }
 
-    NNGraph::TensorNode* embed = embed_tokens_.forward(input_ids);
-    NNGraph::TensorNode* x = transpose(embed, 2);
+    NNGraph::TensorNode* x = embed_tokens_.forward(input_ids);
+    x = transpose(x, 2);
```
Comment (Muxas) [inline nntile/src/model/gptneox/gptneox_model.cc]: Why do we need to transpose after embeddings forward? Seems like we applied embedding layer in an incorrect way. There shall be no transposition after the embed is applied!

File: nntile/include/nntile/model/gpt2/gpt2_mlp.hh [hunk 10]
```diff
@@ -31,19 +31,12 @@ namespace nntile::model::gpt2
...
-    NNGraph::TensorNode* fc1_bias_ = nullptr;
-    NNGraph::TensorNode* fc2_bias_ = nullptr;
-
 public:
     Gpt2MLP(NNGraph* graph,
             const std::string& name,
             const Gpt2Config& config,
             DataType dtype = DataType::FP32);
 
-    NNGraph::TensorNode* forward(
```
Comment (Muxas) [inline nntile/include/nntile/model/gpt2/gpt2_mlp.hh:44]: why is forward operation missing here?

File: nntile/include/nntile/model/gptneox/gptneox_mlp.hh [hunk 11]
```diff
@@ -29,19 +29,12 @@ namespace nntile::model::gptneox
...
-    NNGraph::TensorNode* fc1_bias_ = nullptr;
-    NNGraph::TensorNode* fc2_bias_ = nullptr;
-
 public:
     GptneoxMlp(NNGraph* graph,
                const std::string& name,
                const GptneoxConfig& config,
                DataType dtype = DataType::FP32);
 
-    NNGraph::TensorNode* forward(
```
Comment (Muxas) [inline nntile/include/nntile/model/gptneox/gptneox_mlp.hh:42]: why did you remove this declaration?
